irods/client_configuration/__init__.py

```diff
@@ -0,0 +1,223 @@
+
+class iRODSConfiguration(object):
+    __slots__ = ()
+
+def getter(category, setting):
+    """A programmatic way of allowing the current value of the specified setting to be
+    given indirectly (through an extra call indirection) as the default value of a parameter.
+
+    Returns a lambda that, when called, will yield the setting's value. In the closure of 
+    that lambda, the Python builtin function globals() is used to access (in a read-only
+    capacity) the namespace dict of the irods.client_configuration module.
+
+    See the irods.manager.data_object_manager.DataObjectManager.open(...) method signature
+    for a usage example.
+    """
+    return lambda:getattr(globals()[category], setting)
+
+# #############################################################################
+#
+# Classes for building client configuration categories
+# (irods.client_configuration.data_objects is one such category):
+
+class DataObjects(iRODSConfiguration):
+    __slots__ = ('auto_close',)
+
+    def __init__(self):
+
+        # Setting it in the constructor lets the attribute be a
+        # configurable one and allows a default value of False.
+        #
+        # Running following code will opt in to the the auto-closing
+        # behavior for any data objects opened subsequently.
+        #
+        # >>> import irods.client_configuration as config
+        # >>> irods.client_configuration.data_objects.auto_close = True
+
+        self.auto_close = False
+
+# #############################################################################
+#
+# Instantiations of client-configuration categories:
+
+# The usage "irods.client_configuration.data_objects" reflects the commonly used
+# manager name (session.data_objects) and is thus understood to influence the
+# behavior of data objects.
+#
+# By design, valid configurable targets (e.g. auto_close) are limited to the names
+# listed in the __slots__ member of the category class.
+
+data_objects = DataObjects()
+
+def _var_items(root):
+    if isinstance(root,types.ModuleType):
+        return [(i,v) for i,v in vars(root).items()
+                if isinstance(v,iRODSConfiguration)]
+    if isinstance(root,iRODSConfiguration):
+        return [(i, getattr(root,i)) for i in root.__slots__]
+    return []
+
+def save(root = None, string='', file = ''):
+    """Save the current configuration.
+
+    When called simply as save(), this function simply writes all client settings into
+    a configuration file.
+
+    The 'root' and 'string' parameters are not likely to be overridden when called from an
+    application. They should usually only vary from the default when save() recurses into itself.
+    (However, for due explanation's sake, 'root' specifies at which subtree node to start writing;
+    None denotes the top level.  'string' specifies the dotted prefix and should be empty for
+    an invocation that references the settings' top level namespace, as is true for the trivial
+    save() call.
```

Comment:
Missing the closing parenthesis.

Reply:
Did a slight rewrite in which the paren's are no longer present.

Reply:
Reads fine to me.